### PR TITLE
bpo-20092. Use __index__ in constructors of int, float and complex.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Constructors of `int`, `float` and `complex` will now fall back to `__index__` if corresponding special methods `__int__`, `__float__` and `__complex__` are not defined.

https://bugs.python.org/issue20092

### CODE REVIEW

**Comment by remilapeyre** [comment]
I'm not sure `PyLong_AsDouble` will always raise an exception for the conversion from `int` to `float` like Steven D'Aprano showed in https://mail.python.org/pipermail/python-dev/2019-March/156687.html.

Here the example given by Steven D'Aprano:

```
>>> class Test:
...     def __index__(self):
...             return 2**64 + 1
... 
>>> 2**64 + 1 == int(float(Test()))
False
```

**Comment by serhiy-storchaka** [comment]
Yes, converting integers to float is not always reversible.
```
>>> 2**64 + 1 == int(float(2**64 + 1))
False
```
This is expected behavior.

**Comment by jdemeyer** [comment]
For `__index__`, I prefer the idea of #13106: that will work in all cases where `__int__` is used, not only in those 4 particular constructors.

**Comment by serhiy-storchaka** [comment]
The only cases where `__int__` is used are those particular constructors.

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -779,7 +792,8 @@ are always available.  They are listed here in alphabetical order.
 
    Return an integer object constructed from a number or string *x*, or return
    ``0`` if no arguments are given.  If *x* defines :meth:`__int__`,
-   ``int(x)`` returns ``x.__int__()``.  If *x* defines :meth:`__trunc__`,
+   ``int(x)`` returns ``x.__int__()``.  If *x* defines :meth:`__index__`,
+   it returns ``x.__index__()``.  If *x* defines :meth:`__trunc__`,
```
Comment (mdickinson) [inline Doc/library/functions.rst:797]: There's some question about what the right order should be, but I agree that it makes sense to try `__index__` before `__trunc__` here.

File: Objects/floatobject.c [hunk 2]
```diff
@@ -246,6 +246,15 @@ PyFloat_AsDouble(PyObject *op)
 
     nb = Py_TYPE(op)->tp_as_number;
     if (nb == NULL || nb->nb_float == NULL) {
+        if (nb && nb->nb_index) {
+            PyObject *res = PyNumber_Index(op);
+            if (!res) {
+                return -1;
```
Comment (mdickinson) [inline Objects/floatobject.c:252]: Super-nitpick: I'd prefer `return -1.0;` (leaving the return-type conversion implicit feels odd to me), but I see that we're returning plain-old `1` in the other branch, so I guess this is fine.

**Review by mdickinson — APPROVED** [review]
LGTM: code looks good, and I agree that this is a desirable change.

The one fly in the ointment is that `index` can return something that's not an exact int, which means that there's now another code-path by which `int` can return something that's not of exact type `int`. 

```
>>> class A:
...     def __index__(self): return True
... 
>>> int(A())
<stdin>:1: DeprecationWarning: __index__ returned non-int (type bool).  The ability to return an instance of a strict subclass of int is deprecated, and may be removed in a future version of Python.
1
```

I'd love to finally turn that `DeprecationWarning` into an error, but fear we've run out of time for that for 3.8, and we may want to wait for 3.9 anyway to make sure that everyone's had a chance to react to that warning. (Do you remember which version it got introduced in?)

**Comment by mdickinson** [comment]
Ah, sorry; apparently we _are_ converting to exact int already. (Yay!)

**Comment by mdickinson** [comment]
Hmm: bug found during manual testing:
```
ython 3.8.0a4+ (remotes/serhiy-storchaka/convert-index-to-number:5f980e9716, Jun  1 2019, 19:59:) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class A:
...     def __index__(self): return 2**1024
... 
>>> float(A())
OverflowError: int too large to convert to float

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <class 'float'> returned a result with an error set
```

File: Objects/abstract.c [hunk 3]
```diff
@@ -1480,6 +1487,15 @@ PyNumber_Float(PyObject *o)
         Py_DECREF(res);
         return PyFloat_FromDouble(val);
     }
+    if (m && m->nb_index) {
+        PyObject *res = PyNumber_Index(o);
+        if (!res) {
+            return NULL;
+        }
+        double val = PyLong_AsDouble(res);
```
Comment (mdickinson) [inline Objects/abstract.c:1495]: Missing an error check here.

**Review by mdickinson — CHANGES_REQUESTED** [review]
Updating my approval to "request changes": there's a missing error check in `PyNumber_Float`.

**Review by mdickinson — APPROVED** [review]

**Pull request merged**